### PR TITLE
Get all dataclasses JSON schema tests passing

### PR DESCRIPTION
**Pull request by dmontagu**
Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: pydantic/json_schema.py [hunk 1]
```diff
@@ -674,16 +699,79 @@ def model_schema(self, schema: core_schema.ModelSchema) -> JsonSchemaValue:
...
-        # TODO copy much of model_schema, share logic where possible
-        return {}
+        json_schema = {'type': 'object', 'properties': properties}
+        if required:
+            json_schema['required'] = required
+        return json_schema
+
+    def dataclass_schema(self, schema: core_schema.DataclassSchema) -> JsonSchemaValue:
+        # TODO: Better-share this logic with model_schema
+        #   I'd prefer to clean this up _after_ we rework the approach to customizing dataclass JSON schema though
```
Comment (dmontagu) [inline pydantic/json_schema.py:736]: If we want to clean this up now I'm happy to.

File: pydantic/json_schema.py [hunk 2]
```diff
@@ -674,16 +699,79 @@ def model_schema(self, schema: core_schema.ModelSchema) -> JsonSchemaValue:
...
                     json_schema.setdefault('title', title)
-            # TODO: Should we be setting `allowedProperties: false` if the model's ConfigDict has extra='forbid'?
         return json_schema
 
-    def dataclass_args_schema(self, schema: core_schema.ArgumentsSchema) -> JsonSchemaValue:
-        # TODO copy much of typed_dict_schema, share logic where possible
-        return {}
+    def dataclass_args_schema(self, schema: core_schema.DataclassArgsSchema) -> JsonSchemaValue:
+        # TODO: Better-share this logic with typed_dict_schema
+        #   I'd prefer to clean this up _after_ we rework the approach to customizing dataclass JSON schema though
```
Comment (dmontagu) [inline pydantic/json_schema.py:706]: If we want to clean this up now I'm happy to.

File: pydantic/json_schema.py (hunk 1 above)
Comment (dmontagu) [reply pydantic/json_schema.py:720]: (cleaned this up)

File: pydantic/json_schema.py (hunk 2 above)
Comment (dmontagu) [reply pydantic/json_schema.py]: (cleaned this up)

**Comment by dmontagu** [comment]
I think it would be better to wait to merge this until after #5305, assuming that will be merged in the immediate future.

**Comment by samuelcolvin** [comment]
conflicts, please update.

**Comment by dmontagu** [comment]
please review

File: pydantic/json_schema.py [hunk 3]
```diff
@@ -846,9 +896,15 @@ def field_title_should_be_set(self, schema: CoreSchema | TypedDictField) -> bool
             if override:
                 return self.field_title_should_be_set(override)
 
-            # TODO: This might not be handling some schema types it should
+            # TODO: Fix this next line after merging model changes that put ref on the ModelSchema
```
Comment (samuelcolvin) [inline pydantic/json_schema.py:899]: is this done now?

File: pydantic/json_schema.py [hunk 4]
```diff
@@ -1172,13 +1243,50 @@ def schema(
     return json_schema
 
 
-def model_schema(
-    model: type[BaseModel],
+# TODO: The JSON schema cache has been there from the beginning (https://github.com/pydantic/pydantic/pull/190 —
```
Comment (samuelcolvin) [inline pydantic/json_schema.py:1246]: I think removing it is fine, might be worth moving this comment to the PR body, it'll both make more sense and arguably be more visible there.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

Happy to merge now or fix these comments first.

**Comment by samuelcolvin** [comment]
please update, or merge if you prefer.

**Comment by samuelcolvin** [comment]
awesome, thank you.

**Pull request merged**